vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java

```diff
@@ -101,4 +109,66 @@ public void setAllRowsVisibleProperty_isHeightByRowsAndIsAllRowsVisibleWork() {
         Assert.assertTrue(grid.isHeightByRows());
         Assert.assertTrue(grid.isAllRowsVisible());
     }
+
+    @Test
+    public void dataProviderListeners_gridAttached_oldDataProviderListenerRemoved() {
+        AtomicInteger listenersCount = new AtomicInteger(0);
+        AtomicBoolean oldListenerRemoved = new AtomicBoolean();
+
+        // given
+        Grid<String> grid = new Grid<>();
+        ListDataProvider<String> dataProvider = new ListDataProvider<String>(
+                Collections.emptyList()) {
+            @Override
+            public Registration addDataProviderListener(
+                    DataProviderListener<String> listener) {
+                listenersCount.incrementAndGet();
+                Registration registration = super.addDataProviderListener(
+                        listener);
+                // the first listener is added by Grid in 'setDataProvider'
+                // and it is the one we want to be removed.
+                return listenersCount.get() == 1
+                        ? Registration.combine(registration,
+                                () -> oldListenerRemoved.set(true))
+                        : registration;
+            }
+        };
+
+        // when
+        grid.setDataProvider(dataProvider);
+
+        // then
+        Assert.assertEquals(
+                "Expected exactly two data provider listeners added by "
+                        + "DataCommunicator's and Grid's 'setDataProvider' methods",
+                2, listenersCount.get());
+        Assert.assertFalse(
+                "Expected data provider listener added by Grid not to be "
+                        + "removed before Grid being attached",
+                oldListenerRemoved.get());
+
+        // given
+        listenersCount.set(0);
+
+        // when
+        DataCommunicatorTest.MockUI mockUI = new DataCommunicatorTest.MockUI();
+        mockUI.add(grid);
```

Comment:
As the detach is also removing the listener(s) and it is a critical part of keeping things working, is there another unit test that verifies that detach does it or could you possibly also include verifying that detach cleans up the listeners, so that we can guarantee that the next attach will work ?

Reply:
Yes, good suggestion, I'll check and add if needed.

Reply:
Added